### ISSUE TITLE
OCCT: Memory/State Corruption: Failed Boolean shape operation can cause undefined behaviour until program restart.

### ISSUE THREAD

**Issue by AIRCAP** [FreeCAD/FreeCAD#17159]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Problem description

While playing around with various test cases for Fuzzy on Boolean operations (#17119) I encountered a situation where after a boolean operation caused corrupted geometry, any subsequent boolean corruption (definitely Fuse and Cut) - including PartDesign Pad and Pocket will also create broken geometry when they are created or re-computed, regardless on Fuzzy setting.

The amount of corruption can be subtle (part fails BOP checks when it should have passed them) to crazy kraken (surfaces disappearing or getting incredibly large and warped, entire shapes vanishing)

The only way to get "normal" behaviour back is to close Freecad and re-open it.

It seems like any Boolean operation that "goes wrong" can trigger this, very high (and in case of OCCT 7.3 also too low) fuzzy values are just a convenient way to reproduce the issue.

Steps to reproduce:

1. Build from source in branch #17119 
2. import both brep files from tests/src/Mod/Part/App/brepfiles/
3. make a Part_Fuse of the two parts (cylinder and helix)
4. in Settings Part/PartDesign change FuzzyValue to a very high number ( >10^6) 
5. recompute the Fusion
6. set FuzzyValue back to default (10)
7. recompute the Fusion again

Expected outcome: fusion calculates correctly with sane fuzzy values
Actual outcome: any boolean operations result in weirdness until program restart  - many shapes fail to compute

### Full version info

```shell
OS: Ubuntu 20.04.6 LTS (XFCE/xfce)
Word size of FreeCAD: 64-bit
Version: 1.1.0dev.38925 +21 (Git)
Build type: Debug
Branch: fixBooleans
Hash: e77bcabc730988699b6b57e8744880625f75a2fe
Python 3.8.10, Qt 5.12.8, Coin 4.0.0, Vtk 7.1.1, OCC 7.6.3 (and also OCC 7.3.0)
Locale: English/United States (en_US)
Stylesheet/Theme/QtStyle: unset/FreeCAD Classic/Qt default
```

### Subproject(s) affected?

Core

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by AIRCAP** [comment]
Btw I don't assume there's much we can do if this is a bug deep i OCCT - but it might be good to investigate it to figure out how exactly it can be triggered - and if there's a way to recover (maybe there's a way to "reload" OCCT somehow?) are there any OCCT related permanent memory structures in FreeCAD?  does it help to completely close and reload the document? if not, what other permanent features other than the opened files does FreeCAD have in memory that have OCCT things lingering?

**Comment by AIRCAP** [comment]
ok by the looks of it the reason was that boolean operations can modify the input shapes - if the flag **SetNonDestructive** is not set to true. -  this is of course ireversable unless the entire document is recomputed. the presistence lies in the Shape object of the input Topo Shapes of the boolean, which get modified

most instances of booleans did not set that flag -- fixed in #17119 

note, OCCT prior to 7.1 doesn't implement the flag correctly, we assume no one uses such old OCCT anymore I guess.
